### ISSUE TITLE
Higher CMA footprint due to use of KEY_MAX_* when configuring MediaCodec

### ISSUE THREAD

**Issue by kirin233x** [google/ExoPlayer#8986]
Hi,
Our application integrates with ExoPlayer, but in testing we found that when ExoPlayer and MediaPlayer play the same video source, ExoPlayer consumes 50M more CMA memory than MediaPlayer.  Is there a way to reduce this memory footprint? I understand that both ExoPlayer and MediaPlayer use MediaCodec as the decoder.

**Comment by marcbaechinger** [comment]
It's probably hard to answer without knowing in detail what you are measuring and comparing and without knowing how you have configured the players for your experiment.

When you are talking about CMA footprint, are you referring to the buffers passed to the MediaCodec? Can you clarify a bit how and what you are measuring?

Besides decoder buffers, I think you already know how to configure the buffers sizes with `DefaultLoadControl` when you build ExoPlayer. If you have considered this in your experiment and made sure that both player have the same buffer sizes before comparing, you could try to adjust the buffer sizes and while doing that measure the quality of experience both players provide. These buffer sizes may be a big chunk of the memory footprint when doing media playback and they also contribute to the user experience. In general, knowing what your experiment is doing, like what you measure and how you measure would be helpful for us for giving you a response. 

Have you seen other big chunks of memory usage of ExoPlayer that you think are making the difference?

Besides just comparing these two players, do you have any absolute threshold your app wants to meet? Can you shed some light on the actual requirements you are trying to meet besides 'MediaPlayer uses less'? Do you see increased OOM of your app for some devices or in general?

**Comment by kirin233x** [comment]
Thank you very much for your reply. As for cma memory, we have found a root device, and we can obtain the cma's memory by reading the data from/sys/kernel/debug/cma. However, for some highly performing devices, this value may not change.  We are using a vehicle device that is capable of steadily consuming CMA memory  

According to the analysis of our ROM team, the main use scenario of CMA should be codec.  

Yes.  I modified the DEFAULT_MIN_BUFFER_MS,DEFAULT_MAX_BUFFER_MS,DEFAULT_BUFFER_FOR_PLAYBACK_MS,DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_  REBUFFER_MS,DEFAULT_VIDEO_BUFFER_SIZE,DEFAULT_MIN_BUFFER_SIZE, and DEFAULT_MIN_BUFFER_SIZE in an attempt to verify the impact of cache size, resulting in no reduction in CMA usage.  

I have been a user of ExoPlayer before, but I do not have a deep understanding of its principle. After the analysis of these days, I have not found out any scenes that may have an impact on internal storage except the cache area, which is also the main direction for me to ask for help. Are there any other scenes that will occupy a large amount of memory?  

ExoPlayer and MediaPlayer are compared using the same video source.  The same device. I think my next step should focus on analyzing the differences between the two using MediaCodec.

Thanks again for your reply and forgive my poor English.

**Comment by marcbaechinger** [comment]
If the CMA memory usage is determined by what `MediaCodec` is doing internally, it's difficult to have influence from `ExoPlayer` side I think. 

> compared using the same video source

The buffers that are filled by` ExoPlayer` and passed to the `MediaCodec`, are provided by the `MediaCodec`. It decides about the size of these buffers depends of the format of the video source (the format of the track with best quality in case of adaptive media) that the player declares. In ExoPlayer this happens in `MediaCodecRenderer.initCodec(MediaCodecInfo codecInfo, MediaCrypto crypto)`. It's unclear to me whether `MediaPlayer` does something else here than `ExoPlayer.` I honestly just don't know enough about media player. 

Further, usage of `MediaPlayer` and `ExoPlayer` happens in a context that differs and that may cause the differences you are seeing. For instance, usage of media buffers is in-process for the case of `ExoPlayer`, while it is in another process when using `MediaPlayer`. I don't know what the effect of this on internal memory usage is.

I'm not saying that one or the other player could perform better or not, but it's difficult to measure and explain. A clear indication that one solution is doing better would be seeing more crashes or problems with one or the other solution in an experiment when released and running on users devices.  That's why I was asking if you have some indications regarding this. The larger the number of users in such an experiment is the more a judgment can be done how it behaves.

**Comment by kirin233x** [comment]
Thanks for your reply. Many tests have shown that CODEC is the main factor affecting CMA occupancy.  If you use hard coding, you may use this part of memory.  
Next I'm going to use the soft decoding capabilities provided in ExoPlayer to do video parsing and take a look at the CMA memory footprint.  Thanks.

**Comment by kirin233x** [comment]
After many trials.  I found that one of the major factors affecting CMA memory requests was passing in KEY_MAX_WIDTH and KEY_MAX_HEIGHT when setting the mediaCode.  

About these two attributes, the introduction on the official website is that：
Specify both KEY_MAX_WIDTH and KEY_MAX_HEIGHT to enable adaptive playback (seamless resolution change) for a video decoder that supports it (MediaCodecInfo.CodecCapabilities#FEATURE_AdaptivePlayback). The values are used as hints for the codec: they are the maximum expected resolution to prepare for. Depending on codec support, preparing for larger maximum resolution may require more memory even if that resolution is never reached. These fields have no effect for codecs that do not support adaptive playback.

I have modified the source code directly to remove these two attributes and have not seen any significant effect.  What do you think?

**Comment by marcbaechinger** [comment]
That's what I meant when I say that the Format which ExoPlayer reports determines the amount of memory reserved by the codec. If the maximum resolution is higher the codec needs to reserve more memory even if that is not used when a smaller resolution is chosen by the adaptive algorithm. 

It may be that media player sets another maximum cap than ExoPlayer for the video. Then the codec internally reserves less memory. 

As I mentioned, it is very hard to compare these player's mem consumption because we don't really know whether the players do the same. So if Media player caps the top quality of an adaptive stream then it may use less memory but does not provide the same quality of service.

**Comment by ojw28** [comment]
I'm a bit confused by this issue.

@kirin521 - Is your use case to play _non-adaptive_ content? If so, are the values passed using `KEY_MAX_HEIGHT` and `KEY_MAX_WIDTH` the same values as those passed using `KEY_HEIGHT` and `KEY_WIDTH`, and is that alone sufficient to cause additional memory to be allocated?

If that's the case then we should consider _not_ setting `KEY_MAX_HEIGHT` and `KEY_MAX_WIDTH` for non-adaptive playbacks (we still need them to be set for adaptive playbacks). We'd also want to follow up with the platform team as to whether this is expected behavior.

**Comment by kirin233x** [comment]
Hi, thank you for your attention.  Yes, the KEY_HEIGHT/WIDTH passed to mediaCodec in ExoPlayer is currently set to the same value as the KEY_MAX_HEIGHT/WIDTH.  I'm going to take those two values out, and I don't see any playback problems, and the video cache and all that stuff is working, right.

It is difficult to grab the CMA memory on traditional devices. The problem was a CMA exception on the car. I needed to use a serial port to get the real-time CMA memory footprint.  The chip of the car is RCAR M3. I think the use of CMA memory is most likely related to the codec mode of the chip.  

In the case of KEY_MAX_HEIGHT, a video of 640*480 is played, and the memory of CMA is stable at 120M. After removing this attribute, only 30M memory is needed.  @ojw28

**Comment by ojw28** [comment]
> If that's the case then we should consider not setting KEY_MAX_HEIGHT and KEY_MAX_WIDTH for non-adaptive playbacks (we still need them to be set for adaptive playbacks). We'd also want to follow up with the platform team as to whether this is expected behavior.

@christosts - This seems like a `MediaCodec` question to follow up with the framework team about, probably using an internal bug.

It's unclear why memory allocation would be higher. It might be a platform specific issue, but there may also be a legitimate reason why the underlying implementation would need to do that (e.g., having to allocate a larger number of buffers in anticipation of possible adaptation). If the latter, we should consider populating `KEY_MAX_HEIGHT` and `KEY_MAX_WIDTH` conditionally on adaptive content being played. This is something we can infer from the length of the `streamFormats` array passed to `MediaCodecVideoRenderer.getCodecMaxValues`.

**Comment by christosts** [comment]
internal ref: b/189918386

**Comment by christosts** [comment]
We communicated with the Android platform and they responded that is not the expected behavior with respect to memory allocation. The platform team is not aware of such behavior on devices, and it's possible that this may be root caused to vendor-specific implementation.

The suggestion we can offer is to reach out to the vendor (if possible) or keep the current customization on the ExoPlayer if that solves your problem with memory consumption.

On the ExoPlayer side, we considered not setting the KEY_MAX_WIDTH/HEIGHT for non-adaptive playbacks, but we concluded that this risks any regression on existing playback behavior so we will not move forward with such change.
